src/wallet/rpcwallet.cpp

```diff
@@ -3520,6 +3520,336 @@ UniValue rescanblockchain(const JSONRPCRequest& request)
+            continue;
+        }
+        const CWalletTx& wtx = pwallet->mapWallet.at(txhash);
+        const CTransaction& ctx = *wtx.tx;
+
+        // Get scriptpubkey and check for redeemScript or witnessscript
+        CTxOut prevout = ctx.vout[txin.prevout.n];
+        txnouttype type;
+        std::vector<std::vector<unsigned char>> solns;
+        Solver(prevout.scriptPubKey, type, solns);
+        // Get script hashes
+        if (type == TX_SCRIPTHASH) {
+            // get the hash and find it in the wallet.
+            CScript redeem_script;
+            uint160 hash(solns[0]);
+            pwallet->GetCScript(CScriptID(hash), redeem_script);
+
+            // put redeem_script in map
+            psbtx.redeem_scripts.emplace(hash, redeem_script);
+
+            // Now check whether the redeem_script is a witness script
+            solns.clear();
+            Solver(redeem_script, type, solns);
+        }
+        // Get witness scripts
+        bool witness = false;
+        if (type == TX_WITNESS_V0_SCRIPTHASH) {
+            witness = true;
+            // Get the hash from the solver return
+            uint160 hash;
+            CRIPEMD160().Write(&solns[0][0], solns[0].size()).Finalize(hash.begin());
+
+            // Lookup hash from wallet
+            CScript witness_script;
+            pwallet->GetCScript(CScriptID(hash), witness_script);
+
+            // Put witness script in map
+            uint256 hash256(solns[0]);
+            psbtx.witness_scripts.emplace(hash256, witness_script);
+
+            // Decode the witness script
+            solns.clear();
+            Solver(witness_script, type, solns);
+        }
+        // Get public keys if hd is enabled
+        if (pwallet->IsHDEnabled()) {
+            if (type == TX_PUBKEYHASH) {
+                uint160 hash(solns[0]);
+                CKeyID keyID(hash);
+                add_keypath_to_map(pwallet, keyID, psbtx.hd_keypaths);
+            } else if (type == TX_PUBKEY) {
+                CPubKey vchPubKey(solns[0]);
+                CKeyID keyID = vchPubKey.GetID();
+                add_keypath_to_map(pwallet, keyID, psbtx.hd_keypaths);
+            } else if (type == TX_MULTISIG) {
+                for (auto& soln : solns) {
+                    CPubKey vchPubKey(soln);
+                    CKeyID keyID = vchPubKey.GetID();
+                    add_keypath_to_map(pwallet, keyID, psbtx.hd_keypaths);
+                }
+            }
+        }
+
+        // Put the witness utxo for witness outputs
+        if (witness || type == TX_WITNESS_V0_KEYHASH || type == TX_WITNESS_V0_SCRIPTHASH || type == TX_WITNESS_UNKNOWN) {
+            // Put the witness CTxOut in the input
+            input.witness_utxo = prevout;
+        }
+        // Not witness, put non witness utxo
+        else {
+            input.non_witness_utxo = wtx.tx;
+        }
+
+        if (psbtx_blank) {
+            // Add to inputs
+            psbtx.inputs.push_back(input);
+        }
+    }
```

Comment:
this isn't mentioned in the spec anywhere, but passing along the the change output's pubkey(s)/hdpath(s)/redeemscript allows hww to understand change outputs.

Reply:
Done

Reply:
I decided to make this an option instead of the default. I'll add tests for it later.

Reply:
Should that be mentioned in the spec, then?